pkg/controller/jobs/kubeflow/jobs/pytorchjob/pytorchjob_controller.go

```diff
@@ -0,0 +1,116 @@
+package pytorchjob
+
+import (
+	"context"
+
+	kftraining "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	"sigs.k8s.io/kueue/pkg/controller/jobs/kubeflow/kubeflowjob"
+)
+
+var (
+	gvk           = kftraining.SchemeGroupVersion.WithKind(kftraining.PytorchJobKind)
+	FrameworkName = "kubeflow.org/pytorchjob"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:  SetupIndexes,
+		NewReconciler: NewReconciler,
+		SetupWebhook:  SetupPyTorchJobWebhook,
+		JobType:       &kftraining.PyTorchJob{},
+		AddToScheme:   kftraining.AddToScheme,
+	}))
+}
+
+// +kubebuilder:rbac:groups=scheduling.k8s.io,resources=priorityclasses,verbs=list;get;watch
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update;patch
+// +kubebuilder:rbac:groups=kubeflow.org,resources=pytorchjobs,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=kubeflow.org,resources=pytorchjobs/status,verbs=get;update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/finalizers,verbs=update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=resourceflavors,verbs=get;list;watch
+
+var NewReconciler = jobframework.NewGenericReconciler(func() jobframework.GenericJob {
+	pytorchJobObj := &kftraining.PyTorchJob{}
+	return &PyTorchJob{PyTorchJob: pytorchJobObj, KubeflowJob: kubeflowjob.NewKubeflowJob((*JobControl)(pytorchJobObj))}
+}, nil)
+
+func SetupIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return jobframework.SetupWorkloadOwnerIndex(ctx, indexer, gvk)
+}
+
+func GetWorkloadNameForPyTorchJob(jobName string) string {
+	return jobframework.GetWorkloadNameForOwnerWithGVK(jobName, gvk)
+}
+
+type PyTorchJob struct {
+	*kftraining.PyTorchJob
+	*kubeflowjob.KubeflowJob
+}
```

Comment:
If you make `Object()` mart of the `KFJobControl` you could remove the `kftraining.PyTorchJob` member,  and potentially change this  to :

```suggestion
type PyTorchJob kubeflowjob.KubeflowJob
```

Reply:
It makes sense.

Reply:
For Composition, we need to use an embedded struct. So I will modify this in the following:

```go
type PyTorchJob struct {
	*kubeflowjob.KubeflowJob
}
```

Reply:
I think it should work the same way, it is used like this to "make" the generic job out of all integrations we currently have.

Reply:
@trasc  Does that mean we should use the type alias instead of the embedded struct?

If doing so, it means that we remove `kubeflowjob.KubeflowJob`, and then in all kubeflow integrations (TFJob, PaddleJob, ...), we declare the same processes such as `RunWithPodSetsInfo`, `RestorePodSetsInfo`  and so on.

Reply:
I'd like to use the embedded struct instead of the type alias because I'd like to use method override and then avoid redeclaration of the same processes.

Reply:
The embed approach comes with an extra  allocation every time is passed by value.  Method overwrite should work the same way.  (this is not type alias)

Reply:
Hm. My understanding might not be correct. Could you clarify more about `If you make Object() mart of the KFJobControl`?

Reply:
If Object is part of `KFJobControl`, then you can  change 

```diff
func (j *KubeflowJob) Object() client.Object {
	return j.KFJobControl.Object()
}
```

Then you don't need the `*kftraining.PyTorchJob` memeber. 

Then you can simplify the struct, and also make it a type "redefinition"  `type PyTorchJob kubeflowjob.KubeflowJob`.

Reply:
thinking about it , you could do the same for `GetGVK() ` and might be able to make `KubeflowJob` generic and

```diff
var NewReconciler = jobframework.NewGenericReconciler(func() jobframework.GenericJob {
-	pytorchJobObj := &kftraining.PyTorchJob{}
-	return &PyTorchJob{PyTorchJob: pytorchJobObj, KubeflowJob: kubeflowjob.NewKubeflowJob((*JobControl)(pytorchJobObj))}
+	return &kftraining.PyTorchJob{}
}, nil)

...

-type PyTorchJob struct {
-	*kftraining.PyTorchJob
-	*kubeflowjob.KubeflowJob
-}
+ type PyTorchJob  KubeflowJob[PyTorchJob]
```

Reply:
> If Object is part of KFJobControl, then you can change
> 
> func (j *KubeflowJob) Object() client.Object {
> -	return nil
> +	return j.KFJobControl.Object()
> }
> Then you don't need the *kftraining.PyTorchJob memeber.
> 
> Then you can simplify the struct, and also make it a type "redefinition" type PyTorchJob kubeflowjob.KubeflowJob.

Thanks for the clarifications.
If doing so, I think `type PyTorchJob kubeflowjob.KubeflowJob` doesn't implement the `GenericJob` interface, right?

Although `type JobControl kftraining.PyTorchJob` implement the `GenericJob`.

Reply:
Tooo many interfaces :)... ,  

To keep it simple I think `kubeflowjob.KubeflowJob` should have all the common code for kubeflow(kftraining)) Jobs. and use `KFJobControl` to get it's job  (eg. `kftraining.PyTorchJob`) specific bits.

To avoid confusion,  should be a named member, don't embed.

Reply:
Uhm. Actually, I adopted the same pattern you suggested before submitting this PR. Then, I switched implementation to now one since this one is more reusable.

However, I will roll back my implementation to your suggested one since if @trasc,  who has much understanding of the kueue, is confused by this implementation, then almost everyone will be confused.

Reply:
Done.